services/cd-service/pkg/repository/repository.go

```diff
@@ -173,6 +175,73 @@ func openOrCreate(path string, storageBackend StorageBackend) (*git.Repository,
 	return repo2, err
 }
 
+func GetTags(cfg RepositoryConfig, repoName string, ctx context.Context) (tags []*api.TagData, err error) {
+	repo, err := openOrCreate(repoName, cfg.StorageBackend)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open/create repo: %v", err)
+	}
+
+	var credentials *credentialsStore
+	var certificates *certificateStore
+	if strings.HasPrefix(cfg.URL, "./") || strings.HasPrefix(cfg.URL, "/") {
+	} else {
+		credentials, err = cfg.Credentials.load()
+		if err != nil {
+			return nil, fmt.Errorf("failure to load credentials: %v", err)
+		}
+		certificates, err = cfg.Certificates.load()
+		if err != nil {
+			return nil, fmt.Errorf("failure to load certificates: %v", err)
+		}
+	}
+
+	fetchSpec := fmt.Sprintf("+refs/heads/%s:refs/remotes/origin/%s", cfg.Branch, cfg.Branch)
+	fetchOptions := git.FetchOptions{
+		RemoteCallbacks: git.RemoteCallbacks{
+			CredentialsCallback:      credentials.CredentialsCallback(ctx),
+			CertificateCheckCallback: certificates.CertificateCheckCallback(ctx),
+		},
+		DownloadTags: git.DownloadTagsAll,
+	}
+	remote, err := repo.Remotes.CreateAnonymous(cfg.URL)
+	if err != nil {
+		return nil, fmt.Errorf("failure to create anonymous remote: %v", err)
+	}
+	err = remote.Fetch([]string{fetchSpec}, &fetchOptions, "fetching")
+	if err != nil {
+		return nil, fmt.Errorf("failure to fetch: %v", err)
+	}
+
+	tagsList, err := repo.Tags.List()
+	if err != nil {
+		return nil, fmt.Errorf("unable to list tags: %v", err)
+	}
+
+	sort.Strings(tagsList)
+	iters, err := repo.NewReferenceIteratorGlob("refs/tags/*")
+	if err != nil {
+		return nil, fmt.Errorf("unable to get list of tags: %v", err)
+	}
+	for {
+		tagObject, err := iters.Next()
+		if err != nil {
+			break
+		}
+		tagRef, err := repo.LookupTag(tagObject.Target())
+		if err != nil {
+			tagCommit, err := repo.LookupCommit(tagObject.Target())
+			if err != nil {
+				return nil, fmt.Errorf("unable to lookup tag [%s]: %v", tagObject.Name(), err)
```

Comment:
Please ensure that no error is ever lost:

```suggestion
				return nil, fmt.Errorf("unable to lookup tag [%s]: %v - original err: %v", tagObject.Name(), err, lookupErr)
```